proxy/ParentConsistentHash.cc

```diff
@@ -64,23 +67,127 @@ ParentConsistentHash::getPathHash(HttpRequestData *hrdata, ATSHash64 *h)
+          buffer[slen] = 0;
+          Debug("parent_select", "ParentConsistentHash::%s() Using Over-Ride String='%s'.", __func__, buffer);
+        }
+        h->update(tmp, len);
+        goto done;
+      }
+    }
+  }
 
   // Always hash on '/' because paths returned by ATS are always stripped of it
   h->update("/", 1);
 
   tmp = url->path_get(&len);
-  if (tmp) {
+
+  if (tmp && len > 0) {
+    // Print Original Path
+    if (is_debug_tag_set("parent_select")) {
+      slen = (len > 1023) ? 1023 : len;
+      strncpy(buffer, tmp, slen);
+      buffer[slen] = 0;
+      Debug("parent_select", "ParentConsistentHash::%s() Original Path='%s'.", __func__, buffer);
+    }
+
+    // Process max_dirs directive.
+    if (max_dirs != 0) {
+      // Determine number of directory components in the path.
+      // NOTE: Leading '/' is gone already.
+      for (int x = 0; x < len; x++) {
+        if (tmp[x] == '/')
+          num_dirs++;
+      }
+      // If max_dirs positive , include directory components from the left up to max_dirs.
+      // If max_dirs negative , include directory components from the left up to num_dirs - ( abs(max_dirs) - 1 ).
+      int limit = 0;
+      if (max_dirs > 0)
+        limit = max_dirs;
+      else if (max_dirs < 0) {
+        int md = abs(max_dirs) - 1;
+        if (md < num_dirs)
+          limit = num_dirs - md;
+        else
+          limit = 0;
+      }
+      if (limit > 0) {
+        int x     = 0;
+        int count = 0;
+        for (; x < len; x++) {
+          if (tmp[x] == '/')
+            count++;
+          if (count == limit) {
+            len = x + 1;
+            break;
+          }
+        }
+      } else {
+        len = 0;
+      }
+    }
+
+    // Print Post max_dirs Path
+    if (is_debug_tag_set("parent_select")) {
+      slen = (len > 1023) ? 1023 : len;
+      strncpy(buffer, tmp, slen);
+      buffer[slen] = 0;
+      Debug("parent_select", "ParentConsistentHash::%s() Post-MD Path='%s'.", __func__, buffer);
+    }
+
+    if (ignore_fname || max_dirs != 0) {
+      int x = len - 1;
+      for (; x >= 0; x--) {
+        char c = tmp[x];
+        if (c == '/')
+          break;
+      }
+      len = x + 1;
+    }
+
+    // Print Post fname Path
+    if (is_debug_tag_set("parent_select")) {
+      slen = (len > 1023) ? 1023 : len;
+      strncpy(buffer, tmp, slen);
+      buffer[slen] = 0;
+      Debug("parent_select", "ParentConsistentHash::%s() Post-FN Path='%s'.", __func__, buffer);
+    }
+
     h->update(tmp, len);
   }
 
-  if (!ignore_query) {
+  if (!ignore_query && !ignore_fname && max_dirs == 0) {
     tmp = url->query_get(&len);
     if (tmp) {
       h->update("?", 1);
       h->update(tmp, len);
+      if (is_debug_tag_set("parent_select")) {
+        slen = (len > 1023) ? 1023 : len;
+        strncpy(buffer, tmp, slen);
+        buffer[slen] = 0;
+        Debug("parent_select", "ParentConsistentHash::%s() Added query arguments to hash. Query='%s'.", __func__, buffer);
```

Comment:
Since there are ~4 copies of this, you should wrap it up into a helper function.